packages/message-manager/src/utils.ts

```diff
@@ -98,17 +98,28 @@ export function validateTypedSignMessageDataV3V4(
 ) {
   validateAddress(messageData.from, 'from');
 
-  if (!messageData.data || typeof messageData.data !== 'string') {
+  if (
+    !messageData.data ||
+    Array.isArray(messageData.data) ||
+    (typeof messageData.data !== 'object' &&
+      typeof messageData.data !== 'string')
+  ) {
     throw new Error(
-      `Invalid message "data": ${messageData.data} must be a valid array.`,
+      `Invalid message "data": ${messageData.data} must be a valid string or object.`,
```

Comment:
Nit: Wondering if it might be more useful to explain _what_ constitutes a valid string or object. Like:

``` typescript
      `Invalid message "data": ${messageData.data} must be a valid object or a string that JSON-parses to an object which conforms to the XYZ interface`,
```

Reply:
they both need to conform to [eip-712](https://eips.ethereum.org/EIPS/eip-712) format